plugins/outputs/mqtt/mqtt.go

```diff
@@ -28,61 +26,71 @@ const (
 
 type MQTT struct {
 	Servers     []string `toml:"servers"`
-	Username    string
-	Password    string
+	Protocol    string   `toml:"protocol"`
+	Username    string   `toml:"username"`
+	Password    string   `toml:"password"`
 	Database    string
-	Timeout     config.Duration
-	TopicPrefix string
-	QoS         int    `toml:"qos"`
-	ClientID    string `toml:"client_id"`
+	Timeout     config.Duration `toml:"timeout"`
+	TopicPrefix string          `toml:"topic_prefix"`
+	QoS         int             `toml:"qos"`
+	ClientID    string          `toml:"client_id"`
 	tls.ClientConfig
 	BatchMessage bool            `toml:"batch"`
 	Retain       bool            `toml:"retain"`
 	KeepAlive    int64           `toml:"keep_alive"`
 	Log          telegraf.Logger `toml:"-"`
 
-	client paho.Client
-	opts   *paho.ClientOptions
-
+	client     Client
 	serializer serializers.Serializer
 
 	sync.Mutex
 }
 
+// Client is a protocol neutral MQTT client for connecting,
+// disconnecting, and publishing data to a topic.
+// The protocol specific clients must implement this interface
+type Client interface {
+	Connect() error
+	Publish(topic string, data []byte) error
+	Close() error
+}
+
 func (*MQTT) SampleConfig() string {
 	return sampleConfig
 }
 
 func (m *MQTT) Connect() error {
-	var err error
 	m.Lock()
 	defer m.Unlock()
 	if m.QoS > 2 || m.QoS < 0 {
 		return fmt.Errorf("MQTT Output, invalid QoS value: %d", m.QoS)
 	}
 
-	m.opts, err = m.createOpts()
-	if err != nil {
-		return err
+	if m.KeepAlive == 0 {
+		m.KeepAlive = 30
 	}
```

Comment:
I think it would be better to set `defaultKeepAlive = 30` instead of this indirection.